candid/src/codegen/rust.rs

```diff
@@ -0,0 +1,225 @@
+        unimplemented!()
+    }
+
+    fn usage_service(&self, _ty: &[Binding]) -> Result<String> {
+        unimplemented!()
+    }
+
+    fn declare_prim(&self, id: &str, ty: &PrimType) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_prim(ty)?))
+    }
+    fn declare_var(&self, id: &str, var: &str) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_var(var)?))
+    }
+    fn declare_func(&self, id: &str, func: &FuncType) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_func(func)?))
+    }
+    fn declare_opt(&self, id: &str, ty: &IDLType) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_opt(ty)?))
+    }
+    fn declare_vec(&self, id: &str, ty: &IDLType) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_vec(ty)?))
+    }
+    fn declare_record(&self, id: &str, fields: &[TypeField]) -> Result<String> {
+        let all_fields = fields
+            .iter()
+            .map(|TypeField { label, typ }| {
+                let field_name = match label {
+                    Label::Id(i) => format!("id_{}", i),
+                    Label::Unnamed(i) => format!("id_{}", i),
+                    Label::Named(i) => i.clone(),
+                };
+                let usage = self.usage(typ)?;
+
+                Ok(format!("pub {} : {} ,", field_name, usage))
+            })
+            .collect::<Result<Vec<String>>>()?
+            .join(" ");
+
+        Ok(format!("pub struct {} {{ {} }}", id, all_fields))
+    }
+    fn declare_variant(&self, id: &str, fields: &[TypeField]) -> Result<String> {
+        Ok(format!(
+            "pub type {} = {};",
+            id,
+            self.usage_variant(fields)?
+        ))
+    }
+    fn declare_service(&self, id: &str, ty: &[Binding]) -> Result<String> {
+        Ok(format!("pub type {} = {};", id, self.usage_service(ty)?))
+    }
+
+    fn declaration_binding(&self, binding: &Binding) -> Result<String> {
+        self.declare(&binding.id, &binding.typ)
+    }
+
+    fn service_binding(&self, id: &str, func_t: &FuncType) -> Result<String> {
+        let return_type = if func_t.rets.is_empty() {
+            "()".to_owned()
+        } else {
+            self.usage(&func_t.rets.first().unwrap())?
+        };
+
+        // Add Future binding.
+        let return_type = if func_t.is_query() {
+            return_type
+        } else {
+            format!("impl std::future::Future<Output = {}>", return_type)
+        };
+
+        let arguments = func_t
+            .args
+            .iter()
+            .enumerate()
+            .map(|(i, ty)| {
+                let arg_name = format!("arg{}", i);
+                let ty = self.usage(&ty)?;
+
+                Ok(format!("{}: {}", arg_name, ty))
+            })
+            .collect::<Result<Vec<String>>>()?
+            .join(" , ");
+
+        Ok(format!(
+            "fn {id}( {arguments} ) {return_type} ;",
```

Comment:
Be aware that this `id` can be unicode 🐂

Reply:
That's just illegal... :/

Also, it can be `fn` for example which is a keyword. I need to fix that one too.

Reply:
When importing did from Motoko, we convert the id to `_idl_hash(id)_` when id is not a valid Rust identifier.

Reply:
Done.